File: finForgeUI/src/services/expenseService.js
import apiClient from './apiClient';

/**
 * UI Service: Expense Service
 * Directly interacts with backend API endpoints at /api/expenses
 */
export const expenseService = {
  /**
   * Fetches paginated and optionally filtered expenses from backend
   */
  async getExpenses({ page = 1, pageSize = 10, categoryId, fromDate, toDate } = {}) {
    return await apiClient.get('/expenses', {
      page,
      pageSize,
      categoryId,
      fromDate,
      toDate,
    });
  },

  /**
   * Fetches a single expense record by ID
   */
  async getExpenseById(id) {
    return await apiClient.get(`/expenses/${id}`);
  },

  /**
   * Sends a new expense entry to backend endpoint
   */
  async addExpense(expenseData) {
    return await apiClient.post('/expenses', expenseData);
  },

  /**
   * Updates an existing expense on backend endpoint
   */
  async updateExpense(id, expenseData) {
    return await apiClient.put(`/expenses/${id}`, expenseData);
  },

  /**
   * Deletes an expense from backend
   */
  async deleteExpense(id) {
    return await apiClient.delete(`/expenses/${id}`);
  },
};

export default expenseService;

File: finforge/src/main/java/com/finforge/service/ExpenseService.java
package com.finforge.service;

import com.finforge.dto.ExpenseDTO;
import com.finforge.dto.ExpenseFilterDTO;
import com.finforge.dto.PagedResult;
import com.finforge.exception.DAOException;
import com.finforge.exception.ValidationException;
import com.finforge.model.Expense;

import java.util.List;

/**
 * Service interface for expense management operations.
 */
public interface ExpenseService {

    /** Validates and persists a new expense for the given user. */
    Expense addExpense(int userId, ExpenseDTO dto) throws ValidationException, DAOException;

    /** Validates and updates an existing expense owned by the given user. */
    void updateExpense(int userId, ExpenseDTO dto) throws ValidationException, DAOException;

    /** Deletes an expense by ID after verifying ownership. */
    void deleteExpense(int expenseId, int userId) throws ValidationException, DAOException;

    /** Returns a single expense by ID (must belong to the given user). */
    Expense getExpenseById(int expenseId, int userId) throws ValidationException, DAOException;

    /** Returns all expenses for the given user. */
    List<Expense> getAllExpenses(int userId) throws DAOException;

    /**
     * Returns expenses matching optional date-range and/or category filters.
     * Blank filter fields are ignored.
     */
    List<Expense> searchExpenses(int userId, ExpenseFilterDTO filter)
            throws ValidationException, DAOException;

    /** Returns a page of expenses for the given user, ordered by date descending. */
    PagedResult<Expense> getAllExpensesPaged(int userId, int page, int pageSize) throws DAOException;
}
